Cannot build realm with 5.8.0

#### Goal
<!-- What do you want to achieve? -->
Cannot build realm with 5.8.0

#### Actual Results
Error: Invoke-customs are only supported starting with Android O (--min-api 26)

#### Steps & Code to Reproduce

#### Version of Realm and tooling

Realm version(s): ?
5.8.0

Realm Sync feature enabled: Yes/No

Android Studio version: ?
3.2.1

Android Build Tools version: ?
28

Gradle version: ?
3.2.1

Which Android version and device(s): ?
21

Some more error details in case it helps:
Error: Invoke-customs are only supported starting with Android O (--min-api 26)
Caused by: com.android.tools.r8.utils.AbortException: Error: Invoke-customs are only supported starting with Android O (--min-api 26)
Caused by: com.android.tools.r8.CompilationFailedException: Compilation failed to complete
Caused by: com.android.builder.dexing.DexArchiveBuilderException: Error while dexing.
Caused by: com.android.builder.dexing.DexArchiveBuilderException: Failed to process /Users/mgr/.gradle/caches/transforms-1/files-1.1/realm-android-library-5.8.0.aar/c7f823427f262f54101fbb84c4c4ac4a/jars/classes.jar

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,7 @@
 * None
 
 ### Fixed
-* ?? (Issue [#??](https://github.com/realm/realm-java/issues/??), since ??).
+* Removed Java 8 bytecode. Resulted in errors like `D8: Invoke-customs are only supported starting with Android O (--min-api 26)` if not compiled with Java 8. (Issue [#6300](https://github.com/realm/realm-java/issues/6300), since 5.8.0).
 
 ### Compatibility
 * Realm Object Server: 3.11.0 or later.

--- a/realm/realm-library/src/main/java/io/realm/RealmResults.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmResults.java
@@ -475,7 +475,7 @@ public class RealmResults<E> extends OrderedRealmCollectionImpl<E> {
             case LIST:
                 checkTypeOfListElements(list, RealmModel.class);
                 checkRealmObjectConstraints(fieldName, (RealmModel) list.first(null));
-                osResults.setModelList(fieldName, (RealmList<? extends RealmModel>) list);
+                osResults.setModelList(fieldName, (RealmList<RealmModel>) list);
                 break;
             case INTEGER_LIST:
                 // Integers are a bit annoying as they are all stored as the same type in Core

--- a/realm/realm-library/src/main/java/io/realm/internal/OsResults.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/OsResults.java
@@ -473,68 +473,101 @@ public class OsResults implements NativeObject, ObservableCollection {
     }
 
     public void setStringList(String fieldName, RealmList<String> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addStringList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<String>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<String> list) {
+                builder.addStringList(0, list);
+            }
         });
     }
 
     public void setByteList(String fieldName, RealmList<Byte> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addByteList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Byte>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Byte> list) {
+                builder.addByteList(0, list);
+            }
         });
     }
 
     public void setShortList(String fieldName, RealmList<Short> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addShortList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Short>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Short> list) {
+                builder.addShortList(0, list);
+            }
         });
     }
 
     public void setIntegerList(String fieldName, RealmList<Integer> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addIntegerList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Integer>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Integer> list) {
+                builder.addIntegerList(0, list);
+            }
         });
     }
 
     public void setLongList(String fieldName, RealmList<Long> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addLongList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Long>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Long> list) {
+                builder.addLongList(0, list);
+            }
         });
     }
 
     public void setBooleanList(String fieldName, RealmList<Boolean> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addBooleanList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Boolean>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Boolean> list) {
+                builder.addBooleanList(0, list);
+            }
         });
     }
 
     public void setByteArrayList(String fieldName, RealmList<byte[]> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addByteArrayList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<byte[]>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<byte[]> list) {
+                builder.addByteArrayList(0, list);
+            }
         });
     }
 
     public void setDateList(String fieldName, RealmList<Date> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addDateList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Date>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Date> list) {
+                builder.addDateList(0, list);
+            }
         });
     }
 
     public void setFloatList(String fieldName, RealmList<Float> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addFloatList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Float>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Float> list) {
+                builder.addFloatList(0, list);
+            }
         });
     }
 
     public void setDoubleList(String fieldName, RealmList<Double> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addDoubleList(0, lst);
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<Double>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<Double> list) {
+                builder.addDoubleList(0, list);
+            }
         });
     }
 
-    public void setModelList(String fieldName, RealmList<? extends RealmModel> list) {
-        addTypeSpecificList(fieldName, list, (builder, lst) -> {
-            builder.addObjectList(0, lst);
+    public void setModelList(String fieldName, RealmList<RealmModel> list) {
+        addTypeSpecificList(fieldName, list, new AddListTypeDelegate<RealmModel>() {
+            @Override
+            public void addList(OsObjectBuilder builder, RealmList<RealmModel> list) {
+                builder.addObjectList(0, list);
+            }
         });
     }
 

